File: services/inference_worker/inference_worker/stages/tracker.py
"""Stage 2 -- Multi-object tracking (ByteTrack / OC-SORT).

Assigns persistent track IDs across frames using IoU-based association.
Maintains per-camera TrackState objects with centroid history, age, and
lifecycle (Tentative -> Confirmed -> Deleted).
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING

import numpy as np

from inference_worker.models import Detection, TrackState
from shared_contracts.geometry import BoundingBox, Point2D

if TYPE_CHECKING:
    from inference_worker.settings import TrackerSettings

logger = logging.getLogger(__name__)


def _bbox_to_xyxy(bbox: BoundingBox) -> np.ndarray:
    return np.array([bbox.x, bbox.y, bbox.x + bbox.w, bbox.y + bbox.h])


def _iou_matrix(boxes_a: np.ndarray, boxes_b: np.ndarray) -> np.ndarray:
    """Compute IoU between two sets of xyxy boxes -> (M, N) matrix."""
    if len(boxes_a) == 0 or len(boxes_b) == 0:
        return np.empty((len(boxes_a), len(boxes_b)), dtype=np.float32)

    x1 = np.maximum(boxes_a[:, None, 0], boxes_b[None, :, 0])
    y1 = np.maximum(boxes_a[:, None, 1], boxes_b[None, :, 1])
    x2 = np.minimum(boxes_a[:, None, 2], boxes_b[None, :, 2])
    y2 = np.minimum(boxes_a[:, None, 3], boxes_b[None, :, 3])

    inter = np.maximum(0, x2 - x1) * np.maximum(0, y2 - y1)
    area_a = (boxes_a[:, 2] - boxes_a[:, 0]) * (boxes_a[:, 3] - boxes_a[:, 1])
    area_b = (boxes_b[:, 2] - boxes_b[:, 0]) * (boxes_b[:, 3] - boxes_b[:, 1])
    union = area_a[:, None] + area_b[None, :] - inter

    return inter / (union + 1e-6)


def _linear_assignment(cost_matrix: np.ndarray, threshold: float) -> tuple[list[tuple[int, int]], list[int], list[int]]:
    """Greedy assignment on a cost matrix (higher = better match).

    Returns (matches, unmatched_rows, unmatched_cols).
    """
    matches: list[tuple[int, int]] = []
    matched_rows: set[int] = set()
    matched_cols: set[int] = set()

    if cost_matrix.size == 0:
        return (
            [],
            list(range(cost_matrix.shape[0])),
            list(range(cost_matrix.shape[1])),
        )

    try:
        from scipy.optimize import linear_sum_assignment

        row_idx, col_idx = linear_sum_assignment(-cost_matrix)
        for r, c in zip(row_idx, col_idx):
            if cost_matrix[r, c] >= threshold:
                matches.append((r, c))
                matched_rows.add(r)
                matched_cols.add(c)
    except ImportError:
        flat = cost_matrix.flatten()
        sorted_indices = np.argsort(-flat)
        for idx in sorted_indices:
            r, c = divmod(int(idx), cost_matrix.shape[1])
            if r in matched_rows or c in matched_cols:
                continue
            if cost_matrix[r, c] < threshold:
                break
            matches.append((r, c))
            matched_rows.add(r)
            matched_cols.add(c)

    unmatched_rows = [i for i in range(cost_matrix.shape[0]) if i not in matched_rows]
    unmatched_cols = [j for j in range(cost_matrix.shape[1]) if j not in matched_cols]
    return matches, unmatched_rows, unmatched_cols


class ByteTracker:
    """ByteTrack-style multi-object tracker.

    Two-stage association: first match high-confidence detections to existing
    tracks, then attempt to recover unmatched tracks using low-confidence
    detections.
    """

    def __init__(self, settings: TrackerSettings, camera_id: str) -> None:
        self._settings = settings
        self._camera_id = camera_id
        self._next_id = 0

    def _new_track_id(self) -> str:
        tid = f"trk_{self._next_id:05d}"
        self._next_id += 1
        return tid

    def update(
        self,
        detections: list[Detection],
        existing_tracks: dict[str, TrackState],
        frame_index: int,
        fps: float = 10.0,
    ) -> dict[str, TrackState]:
        """Run one tracking step.

        Returns the updated track state dict (may add new tracks, remove dead ones).
        """
        settings = self._settings

        high_conf = [d for d in detections if d.confidence >= 0.5]
        low_conf = [d for d in detections if d.confidence < 0.5]

        confirmed = {k: v for k, v in existing_tracks.items() if v.is_confirmed}
        tentative = {k: v for k, v in existing_tracks.items() if not v.is_confirmed}

        updated: dict[str, TrackState] = {}

        matched_track_ids, remaining_high = self._associate(
            confirmed, high_conf, frame_index, fps, updated
        )

        unmatched_confirmed = {
            k: v for k, v in confirmed.items() if k not in matched_track_ids
        }
        self._associate(
            unmatched_confirmed, low_conf, frame_index, fps, updated
        )

        still_unmatched = {
            k: v for k, v in unmatched_confirmed.items() if k not in updated
        }
        for tid, ts in still_unmatched.items():
            ts.time_since_update += 1
            ts.occlusion_flag = True
            if ts.time_since_update <= settings.max_age:
                updated[tid] = ts

        tent_matched, _ = self._associate(
            tentative, remaining_high, frame_index, fps, updated
        )
        unmatched_tentative = {
            k: v for k, v in tentative.items() if k not in tent_matched and k not in updated
        }
        for tid, ts in unmatched_tentative.items():
            ts.time_since_update += 1
            if ts.time_since_update <= settings.max_age:
                updated[tid] = ts

        all_remaining_dets = [d for d in remaining_high if d not in []]
        matched_det_indices: set[int] = set()
        for d_idx, det in enumerate(remaining_high):
            already_used = False
            for ts in updated.values():
                if ts.last_seen_frame == frame_index and ts.bbox == det.bbox:
                    already_used = True
                    break
            if already_used:
                matched_det_indices.add(d_idx)

        for d_idx, det in enumerate(remaining_high):
            if d_idx in matched_det_indices:
                continue
            new_id = self._new_track_id()
            centroid = det.bbox.center
            updated[new_id] = TrackState(
                track_id=new_id,
                bbox=det.bbox,
                centroid=centroid,
                centroid_history=[centroid],
                first_seen_frame=frame_index,
                last_seen_frame=frame_index,
                age=1,
                hits=1,
                time_since_update=0,
                is_confirmed=False,
            )

        for ts in updated.values():
            if not ts.is_confirmed and ts.hits >= settings.min_hits:
                ts.is_confirmed = True

        return updated

    def _associate(
        self,
        tracks: dict[str, TrackState],
        detections: list[Detection],
        frame_index: int,
        fps: float,
        output: dict[str, TrackState],
    ) -> tuple[set[str], list[Detection]]:
        """Associate detections to tracks via IoU, update matched tracks.

        Returns (matched_track_ids, unmatched_detections).
        """
        if not tracks or not detections:
            return set(), list(detections)

        track_ids = list(tracks.keys())
        track_boxes = np.array([_bbox_to_xyxy(tracks[tid].bbox) for tid in track_ids])
        det_boxes = np.array([_bbox_to_xyxy(d.bbox) for d in detections])

        iou = _iou_matrix(track_boxes, det_boxes)
        matches, unmatched_t, unmatched_d = _linear_assignment(
            iou, self._settings.iou_threshold
        )

        matched_tids: set[str] = set()
        matched_dets: set[int] = set()

        for t_idx, d_idx in matches:
            tid = track_ids[t_idx]
            det = detections[d_idx]
            ts = tracks[tid]

            centroid = det.bbox.center
            ts.bbox = det.bbox
            ts.centroid = centroid
            ts.centroid_history.append(centroid)
            if len(ts.centroid_history) > self._settings.centroid_history_length:
                ts.centroid_history = ts.centroid_history[-self._settings.centroid_history_length :]
            ts.last_seen_frame = frame_index
            ts.age += 1
            ts.hits += 1
            ts.time_since_update = 0
            ts.occlusion_flag = False

            if len(ts.centroid_history) >= 2 and fps > 0:
                prev = ts.centroid_history[-2]
                disp = ((centroid.x - prev.x) ** 2 + (centroid.y - prev.y) ** 2) ** 0.5
                ts.speed_estimate_kmh = disp * fps * 3.6

            output[tid] = ts
            matched_tids.add(tid)
            matched_dets.add(d_idx)

        remaining = [d for i, d in enumerate(detections) if i not in matched_dets]
        return matched_tids, remaining

    def set_next_id(self, next_id: int) -> None:
        self._next_id = next_id

File: apps/mobile_flutter/lib/core/inference/tracker.dart
/// Stage 2 — ByteTrack-style multi-object tracker.
///
/// Assigns persistent track IDs across frames using IoU-based association.
/// Two-stage matching: first high-confidence detections against confirmed
/// tracks, then low-confidence detections to recover unmatched tracks.
///
/// Ported from `services/inference_worker/inference_worker/stages/tracker.py`.
library;

import 'dart:math' as math;

import 'package:greyeye_mobile/core/inference/models.dart';
import 'package:greyeye_mobile/core/inference/pipeline_settings.dart';

class ByteTracker {
  ByteTracker(this._settings);

  final TrackerSettings _settings;
  int _nextId = 0;

  String _newTrackId() {
    final tid = 'trk_${_nextId.toString().padLeft(5, '0')}';
    _nextId++;
    return tid;
  }

  /// Run one tracking step, returning the updated track-state map.
  Map<String, TrackState> update(
    List<Detection> detections,
    Map<String, TrackState> existingTracks,
    int frameIndex, {
    double fps = 10.0,
  }) {
    final highConf = detections.where((d) => d.confidence >= 0.5).toList();
    final lowConf = detections.where((d) => d.confidence < 0.5).toList();

    final confirmed = <String, TrackState>{};
    final tentative = <String, TrackState>{};
    for (final e in existingTracks.entries) {
      if (e.value.isConfirmed) {
        confirmed[e.key] = e.value;
      } else {
        tentative[e.key] = e.value;
      }
    }

    final updated = <String, TrackState>{};

    // First association: confirmed tracks ↔ high-confidence detections.
    final (matchedTids1, remainingHigh) =
        _associate(confirmed, highConf, frameIndex, fps, updated);

    // Second association: unmatched confirmed ↔ low-confidence detections.
    final unmatchedConfirmed = <String, TrackState>{
      for (final e in confirmed.entries)
        if (!matchedTids1.contains(e.key)) e.key: e.value,
    };
    _associate(unmatchedConfirmed, lowConf, frameIndex, fps, updated);

    // Age out still-unmatched confirmed tracks.
    for (final e in unmatchedConfirmed.entries) {
      if (updated.containsKey(e.key)) continue;
      final ts = e.value;
      ts.timeSinceUpdate += 1;
      ts.occlusionFlag = true;
      if (ts.timeSinceUpdate <= _settings.maxAge) {
        updated[e.key] = ts;
      }
    }

    // Third association: tentative tracks ↔ remaining high-confidence.
    final (tentMatched, _) =
        _associate(tentative, remainingHigh, frameIndex, fps, updated);

    // Age out unmatched tentative tracks.
    for (final e in tentative.entries) {
      if (tentMatched.contains(e.key) || updated.containsKey(e.key)) continue;
      final ts = e.value;
      ts.timeSinceUpdate += 1;
      if (ts.timeSinceUpdate <= _settings.maxAge) {
        updated[e.key] = ts;
      }
    }

    // Create new tentative tracks from unmatched high-confidence detections.
    final usedDetIndices = <int>{};
    for (var i = 0; i < remainingHigh.length; i++) {
      final det = remainingHigh[i];
      for (final ts in updated.values) {
        if (ts.lastSeenFrame == frameIndex &&
            ts.bbox.x == det.bbox.x &&
            ts.bbox.y == det.bbox.y) {
          usedDetIndices.add(i);
          break;
        }
      }
    }

    for (var i = 0; i < remainingHigh.length; i++) {
      if (usedDetIndices.contains(i)) continue;
      final det = remainingHigh[i];
      final newId = _newTrackId();
      final centroid = bboxCenter(det.bbox);
      updated[newId] = TrackState(
        trackId: newId,
        bbox: det.bbox,
        centroid: centroid,
        centroidHistory: [centroid],
        firstSeenFrame: frameIndex,
        lastSeenFrame: frameIndex,
        age: 1,
        hits: 1,
      );
    }

    // Promote tentative → confirmed.
    for (final ts in updated.values) {
      if (!ts.isConfirmed && ts.hits >= _settings.minHits) {
        ts.isConfirmed = true;
      }
    }

    return updated;
  }

  /// Associate detections to tracks via IoU.
  ///
  /// Returns (matched track IDs, unmatched detections).
  (Set<String>, List<Detection>) _associate(
    Map<String, TrackState> tracks,
    List<Detection> detections,
    int frameIndex,
    double fps,
    Map<String, TrackState> output,
  ) {
    if (tracks.isEmpty || detections.isEmpty) {
      return ({}, List<Detection>.of(detections));
    }

    final trackIds = tracks.keys.toList();
    final trackBoxes = trackIds.map((id) => bboxToXyxy(tracks[id]!.bbox)).toList();
    final detBoxes = detections.map((d) => bboxToXyxy(d.bbox)).toList();

    final iou = _iouMatrix(trackBoxes, detBoxes);
    final (matches, _, _) = _greedyAssignment(iou, _settings.iouThreshold);

    final matchedTids = <String>{};
    final matchedDetIdx = <int>{};

    for (final (tIdx, dIdx) in matches) {
      final tid = trackIds[tIdx];
      final det = detections[dIdx];
      final ts = tracks[tid]!;

      final centroid = bboxCenter(det.bbox);
      ts.bbox = det.bbox;
      ts.centroid = centroid;
      ts.centroidHistory.add(centroid);
      if (ts.centroidHistory.length > _settings.centroidHistoryLength) {
        ts.centroidHistory = ts.centroidHistory.sublist(
          ts.centroidHistory.length - _settings.centroidHistoryLength,
        );
      }
      ts.lastSeenFrame = frameIndex;
      ts.age += 1;
      ts.hits += 1;
      ts.timeSinceUpdate = 0;
      ts.occlusionFlag = false;

      if (ts.centroidHistory.length >= 2 && fps > 0) {
        final prev = ts.centroidHistory[ts.centroidHistory.length - 2];
        final disp = math.sqrt(
          math.pow(centroid.x - prev.x, 2) + math.pow(centroid.y - prev.y, 2),
        );
        ts.speedEstimateKmh = disp * fps * 3.6;
      }

      output[tid] = ts;
      matchedTids.add(tid);
      matchedDetIdx.add(dIdx);
    }

    final remaining = <Detection>[
      for (var i = 0; i < detections.length; i++)
        if (!matchedDetIdx.contains(i)) detections[i],
    ];

    return (matchedTids, remaining);
  }
}

/// Compute IoU between two sets of xyxy boxes → (M × N) matrix.
List<List<double>> _iouMatrix(
  List<List<double>> boxesA,
  List<List<double>> boxesB,
) {
  final m = boxesA.length;
  final n = boxesB.length;
  final result = List.generate(m, (_) => List<double>.filled(n, 0.0));

  for (var i = 0; i < m; i++) {
    final a = boxesA[i];
    final areaA = (a[2] - a[0]) * (a[3] - a[1]);
    for (var j = 0; j < n; j++) {
      final b = boxesB[j];
      final xx1 = math.max(a[0], b[0]);
      final yy1 = math.max(a[1], b[1]);
      final xx2 = math.min(a[2], b[2]);
      final yy2 = math.min(a[3], b[3]);
      final inter = math.max(0.0, xx2 - xx1) * math.max(0.0, yy2 - yy1);
      final areaB = (b[2] - b[0]) * (b[3] - b[1]);
      result[i][j] = inter / (areaA + areaB - inter + 1e-6);
    }
  }

  return result;
}

/// Greedy assignment on an IoU cost matrix (higher = better).
(List<(int, int)>, List<int>, List<int>) _greedyAssignment(
  List<List<double>> costMatrix,
  double threshold,
) {
  final m = costMatrix.length;
  if (m == 0) return ([], [], []);
  final n = costMatrix[0].length;
  if (n == 0) return ([], List<int>.generate(m, (i) => i), []);

  final entries = <(int, int, double)>[];
  for (var i = 0; i < m; i++) {
    for (var j = 0; j < n; j++) {
      entries.add((i, j, costMatrix[i][j]));
    }
  }
  entries.sort((a, b) => b.$3.compareTo(a.$3));

  final matchedRows = <int>{};
  final matchedCols = <int>{};
  final matches = <(int, int)>[];

  for (final (r, c, cost) in entries) {
    if (matchedRows.contains(r) || matchedCols.contains(c)) continue;
    if (cost < threshold) break;
    matches.add((r, c));
    matchedRows.add(r);
    matchedCols.add(c);
  }

  final unmatchedRows = [
    for (var i = 0; i < m; i++)
      if (!matchedRows.contains(i)) i,
  ];
  final unmatchedCols = [
    for (var j = 0; j < n; j++)
      if (!matchedCols.contains(j)) j,
  ];

  return (matches, unmatchedRows, unmatchedCols);
}
